### PR TITLE
Azure Functions Typescript, Unit Testing, Integration Testing

### PR DESCRIPTION
**Pull request by swissarmykirpan**
A Work in Progress. Some thoughts about how to implement azure functions using typescript, with repositories and testing. 

I've removed the serverless config for now - because I think we would be better off using Terraform and Azure Resource Manager Templates, but that's up for discussion.

function handler -> repository -> storage

This uses azure table storage to store entities for now, but I am seriously considering whether we should switch to using a graph database, such as Gremlin (Apache Tinkerpop) for which there is support in Azure Cosmos DB

Apologies it has taken me so long to get this up here. This is by no means near the finished article and hopefully will provide some direction into my thoughts into how to use this.

### CODE REVIEW

File: add-mentee/index.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { AzureFunction } from "@azure/functions";
+import { container } from "tsyringe";
+import "reflect-metadata";
+import { setupContainer } from "../container";
+import { AddMentee } from "./add-mentee-handler";
+
+setupContainer();
+
```
Comment (swissarmykirpan) [inline add-mentee/index.ts]: current work around - but will be replaced shortly since the new TS AzureFunctions helper has been released

**Comment by swissarmykirpan** [comment]
@crysfel @jagzviruz

**Comment by crysfel** [comment]
Thanks for contributing! I have two silly questions:

1.- I've installed everything in my local and now it seems to run successfully. My question now is, how can I find out which URLs I can use to execute the functions?

2.- How can I deploy this to azure and see it live?

**Comment by swissarmykirpan** [comment]
This isn't ready to execute (just yet), so please bear with me.
1. to run it locally, you should just be able to run yarn start. this will bring up a terminal and tell you which endpoints it has found. If you look in the source code you will find `add-mentee-handler`- that's the code.   
2. that's yet to come. I need to create azure resource manager templates etc - there's a fair amount of work to create environments yet to come. - this is partly why i wanted front-end and back-end to exist in a single monorepo - so there was no need on infrastructure to see things work - while all of this gets setup

**Comment by crysfel** [comment]
That's what I did, but I don't see any URL I can hit using postman. Here's the logs I'm seeing after running `yarn start`

```
$ yarn start
yarn run v1.3.2
(node:71560) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
warning package.json: No license field
$ yarn build && func extensions install
(node:71562) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
warning package.json: No license field
$ tsc
Microsoft (R) Build Engine version 15.9.20+g88f5fadfbe for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 194 ms for /Users/crysfel/Workspace/codingcoach/backend/extensions.csproj.
  extensions -> /Users/crysfel/Workspace/codingcoach/backend/bin/extensions.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.68


$ yarn start:host & yarn watch
(node:71574) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(node:71575) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
warning package.json: No license field
warning package.json: No license field
$ func start
$ tsc -w
[9:18:47 AM] Starting compilation in watch mode...

your worker runtime is not set. As of 2.0.1-beta.26 a worker runtime setting is required.
Please run `func settings add FUNCTIONS_WORKER_RUNTIME <option>` or add FUNCTIONS_WORKER_RUNTIME to your local.settings.json
Available options: dotnet, node, python

                  %%%%%%
                 %%%%%%
            @   %%%%%%    @
          @@   %%%%%%      @@
       @@@    %%%%%%%%%%%    @@@
     @@      %%%%%%%%%%        @@
       @@         %%%%       @@
         @@      %%%       @@
           @@    %%      @@
                %%
                %

Azure Functions Core Tools (2.5.553 Commit hash: b63f2d6f5479718ca276f7b585a8cc6b9bc57c4f)
Function Runtime Version: 2.0.12382.0
[4/10/19 2:18:50 PM] Starting Rpc Initialization Service.
[4/10/19 2:18:50 PM] Initializing RpcServer
[4/10/19 2:18:50 PM] Building host: startup suppressed:False, configuration suppressed: False
[4/10/19 2:18:50 PM] Reading host configuration file '/Users/crysfel/Workspace/codingcoach/backend/host.json'
[4/10/19 2:18:50 PM] Host configuration file read:
[4/10/19 2:18:50 PM] {
[4/10/19 2:18:50 PM]   "version": "2.0"
[4/10/19 2:18:50 PM] }
[4/10/19 2:18:51 PM] Initializing Host.
[4/10/19 2:18:51 PM] Host initialization: ConsecutiveErrors=0, StartupCount=1
[4/10/19 2:18:51 PM] LoggerFilterOptions
[4/10/19 2:18:51 PM] {
[4/10/19 2:18:51 PM]   "MinLevel": "None",
[4/10/19 2:18:51 PM]   "Rules": [
[4/10/19 2:18:51 PM]     {
[4/10/19 2:18:51 PM]       "ProviderName": null,
[4/10/19 2:18:51 PM]       "CategoryName": null,
[4/10/19 2:18:51 PM]       "LogLevel": null,
[4/10/19 2:18:51 PM]       "Filter": "<AddFilter>b__0"
[4/10/19 2:18:51 PM]     },
[4/10/19 2:18:51 PM]     {
[4/10/19 2:18:51 PM]       "ProviderName": "Microsoft.Azure.WebJobs.Script.WebHost.Diagnostics.SystemLoggerProvider",
[4/10/19 2:18:51 PM]       "CategoryName": null,
[4/10/19 2:18:51 PM]       "LogLevel": "None",
[4/10/19 2:18:51 PM]       "Filter": null
[4/10/19 2:18:51 PM]     },
[4/10/19 2:18:51 PM]     {
[4/10/19 2:18:51 PM]       "ProviderName": "Microsoft.Azure.WebJobs.Script.WebHost.Diagnostics.SystemLoggerProvider",
[4/10/19 2:18:51 PM]       "CategoryName": null,
[4/10/19 2:18:51 PM]       "LogLevel": null,
[4/10/19 2:18:51 PM]       "Filter": "<AddFilter>b__0"
[4/10/19 2:18:51 PM]     }
[4/10/19 2:18:51 PM]   ]
[4/10/19 2:18:51 PM] }
[4/10/19 2:18:51 PM] FunctionResultAggregatorOptions
[4/10/19 2:18:51 PM] {
[4/10/19 2:18:51 PM]   "BatchSize": 1000,
[4/10/19 2:18:51 PM]   "FlushTimeout": "00:00:30",
[4/10/19 2:18:51 PM]   "IsEnabled": true
[4/10/19 2:18:51 PM] }
[4/10/19 2:18:51 PM] SingletonOptions
[4/10/19 2:18:51 PM] {
[4/10/19 2:18:51 PM]   "LockPeriod": "00:00:15",
[4/10/19 2:18:51 PM]   "ListenerLockPeriod": "00:00:15",
[4/10/19 2:18:51 PM]   "LockAcquisitionTimeout": "10675199.02:48:05.4775807",
[4/10/19 2:18:51 PM]   "LockAcquisitionPollingInterval": "00:00:05",
[4/10/19 2:18:51 PM]   "ListenerLockRecoveryPollingInterval": "00:01:00"
[4/10/19 2:18:51 PM] }
[4/10/19 2:18:51 PM] Starting JobHost
[4/10/19 2:18:51 PM] Starting Host (HostId=crysfelsmacbookpro-880132143, InstanceId=ba8aaf97-ae23-4601-8705-eaf14bb76281, Version=2.0.12382.0, ProcessId=71576, AppDomainId=1, InDebugMode=False, InDiagnosticMode=False, FunctionsExtensionVersion=)
[4/10/19 2:18:51 PM] Loading functions metadata
[4/10/19 2:18:51 PM] 0 functions loaded
[4/10/19 2:18:51 PM] Loading proxies metadata
[4/10/19 2:18:51 PM] Initializing Azure Function proxies
[9:18:51 AM] Found 0 errors. Watching for file changes.

[4/10/19 2:18:51 PM] 0 proxies loaded
[4/10/19 2:18:51 PM] WorkerRuntime: (null). Will shutdown other standby channels
[4/10/19 2:18:51 PM] Generating 0 job function(s)
[4/10/19 2:18:51 PM] No job functions found. Try making your job classes and methods public. If you're using binding extensions (e.g. Azure Storage, ServiceBus, Timers, etc.) make sure you've called the registration method for the extension(s) in your startup code (e.g. builder.AddAzureStorage(), builder.AddServiceBus(), builder.AddTimers(), etc.).
[4/10/19 2:18:51 PM] Host initialized (730ms)
[4/10/19 2:18:51 PM] Host started (747ms)
[4/10/19 2:18:51 PM] Job host started
Hosting environment: Production
Content root path: /Users/crysfel/Workspace/codingcoach/backend
Now listening on: http://0.0.0.0:7071
Application started. Press Ctrl+C to shut down.
[4/10/19 2:18:56 PM] Host lock lease acquired by instance ID '000000000000000000000000448767BF'.
```

Maybe my understanding is wrong and there's no URL available?

**Comment by crysfel** [comment]
My bad, I closed it by mistake 😅

File: package.json [hunk 2]
```diff
@@ -1,13 +1,47 @@
...
+    "build:configFiles": "copyfiles -u 1 \"$npm_package_config_azureFunctions_rootDir/**/*.json\" \"$npm_package_config_azureFunctions_rootDir/**/*.csproj\" $npm_package_config_azureFunctions_outDir",
+    "cdout": "cd $npm_package_config_azureFunctions_outDir && pwd",
+    "build:extensions": "func extensions sync --prefix $npm_package_config_azureFunctions_outDir --output bin",
+    "install:extensions": "yarn build && func extensions install --prefix $npm_package_config_azureFunctions_outDir && copyfiles -u 1 $npm_package_config_azureFunctions_outDir/extensions.csproj $npm_package_config_azureFunctions_rootDir",
+    "install:extensions:force": "yarn build && func extensions install --prefix $npm_package_config_azureFunctions_outDir --force && copyfiles -u 1 $npm_package_config_azureFunctions_outDir/extensions.csproj $npm_package_config_azureFunctions_rootDir",
+    "start": "yarn build || yarn start:host & yarn watch",
```
Comment (crysfel) [inline package.json]: I've been trying to start the server, but it only works if I use an `&&` instead of `||`.

Comment (swissarmykirpan) [reply package.json]: yup, i've made this change

**Pull request merged**